File: specs/001-multi-language-canvas/samples/sample.py
"""
Sample Python file for testing Living Canvas multi-language support.

This file contains various function definitions and call patterns
to test the tree-sitter based call graph extraction.
"""


def greet(name):
    """Return a greeting message."""
    return f"Hello, {name}!"


def farewell(name):
    """Return a farewell message."""
    return f"Goodbye, {name}!"


def process_name(name):
    """Process a name by greeting and then saying farewell."""
    greeting = greet(name)
    goodbye = farewell(name)
    return f"{greeting} ... {goodbye}"


def validate_input(data):
    """Validate input data."""
    if not data:
        return False
    if not isinstance(data, dict):
        return False
    return True


def transform_data(data):
    """Transform data into a new format."""
    if not validate_input(data):
        return None
    return {k.upper(): v for k, v in data.items()}


def save_result(result):
    """Save result to storage (mock)."""
    print(f"Saving: {result}")
    return True


def process_pipeline(data):
    """Main processing pipeline."""
    if not validate_input(data):
        print("Invalid input")
        return None

    transformed = transform_data(data)
    if transformed:
        save_result(transformed)
    return transformed


class Calculator:
    """Simple calculator class."""

    def __init__(self, initial_value=0):
        """Initialize calculator with a value."""
        self.value = initial_value

    def add(self, x):
        """Add x to current value."""
        self.value += x
        return self

    def subtract(self, x):
        """Subtract x from current value."""
        self.value -= x
        return self

    def multiply(self, x):
        """Multiply current value by x."""
        self.value *= x
        return self

    def divide(self, x):
        """Divide current value by x."""
        if x != 0:
            self.value /= x
        return self

    def reset(self):
        """Reset value to zero."""
        self.value = 0
        return self

    def get_value(self):
        """Get current value."""
        return self.value


def use_calculator():
    """Demonstrate calculator usage."""
    calc = Calculator(10)
    calc.add(5).multiply(2).subtract(3)
    return calc.get_value()


def fibonacci(n):
    """Calculate nth Fibonacci number recursively."""
    if n <= 1:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)


def factorial(n):
    """Calculate factorial of n."""
    if n <= 1:
        return 1
    return n * factorial(n - 1)


def math_operations(x, y):
    """Perform various math operations."""
    fib_x = fibonacci(x)
    fact_y = factorial(y)
    return fib_x + fact_y


def main():
    """Main entry point."""
    # Test greeting functions
    message = process_name("World")
    print(message)

    # Test data pipeline
    data = {"name": "test", "value": 42}
    result = process_pipeline(data)
    print(f"Pipeline result: {result}")

    # Test calculator
    calc_result = use_calculator()
    print(f"Calculator result: {calc_result}")

    # Test math operations
    math_result = math_operations(10, 5)
    print(f"Math result: {math_result}")


if __name__ == "__main__":
    main()

File: specs/001-multi-language-canvas/samples/sample.js
/**
 * Sample JavaScript file for testing Living Canvas multi-language support.
 *
 * This file contains various function patterns to test tree-sitter
 * based call graph extraction.
 */

// Regular function declarations
function greet(name) {
    return `Hello, ${name}!`;
}

function farewell(name) {
    return `Goodbye, ${name}!`;
}

function processName(name) {
    const greeting = greet(name);
    const goodbye = farewell(name);
    return `${greeting} ... ${goodbye}`;
}

// Arrow functions
const validateInput = (data) => {
    if (!data) return false;
    if (typeof data !== 'object') return false;
    return true;
};

const transformData = (data) => {
    if (!validateInput(data)) return null;
    return Object.fromEntries(
        Object.entries(data).map(([k, v]) => [k.toUpperCase(), v])
    );
};

const saveResult = (result) => {
    console.log('Saving:', result);
    return true;
};

// Mixed function styles
function processPipeline(data) {
    if (!validateInput(data)) {
        console.log('Invalid input');
        return null;
    }

    const transformed = transformData(data);
    if (transformed) {
        saveResult(transformed);
    }
    return transformed;
}

// Class with methods
class Calculator {
    constructor(initialValue = 0) {
        this.value = initialValue;
    }

    add(x) {
        this.value += x;
        return this;
    }

    subtract(x) {
        this.value -= x;
        return this;
    }

    multiply(x) {
        this.value *= x;
        return this;
    }

    divide(x) {
        if (x !== 0) {
            this.value /= x;
        }
        return this;
    }

    reset() {
        this.value = 0;
        return this;
    }

    getValue() {
        return this.value;
    }
}

function useCalculator() {
    const calc = new Calculator(10);
    calc.add(5).multiply(2).subtract(3);
    return calc.getValue();
}

// Recursive functions
function fibonacci(n) {
    if (n <= 1) return n;
    return fibonacci(n - 1) + fibonacci(n - 2);
}

function factorial(n) {
    if (n <= 1) return 1;
    return n * factorial(n - 1);
}

const mathOperations = (x, y) => {
    const fibX = fibonacci(x);
    const factY = factorial(y);
    return fibX + factY;
};

// Main entry point
function main() {
    // Test greeting functions
    const message = processName('World');
    console.log(message);

    // Test data pipeline
    const data = { name: 'test', value: 42 };
    const result = processPipeline(data);
    console.log('Pipeline result:', result);

    // Test calculator
    const calcResult = useCalculator();
    console.log('Calculator result:', calcResult);

    // Test math operations
    const mathResult = mathOperations(10, 5);
    console.log('Math result:', mathResult);
}

// Export for module usage
if (typeof module !== 'undefined') {
    module.exports = {
        greet,
        farewell,
        processName,
        validateInput,
        transformData,
        processPipeline,
        Calculator,
        useCalculator,
        fibonacci,
        factorial,
        mathOperations,
        main
    };
}

main();
